### ISSUE TITLE
High-level plan for sender adaptation

### ISSUE THREAD

**Issue by msimberg** [eth-cscs/DLA-Future#437]
After #371 was merged the basics for using senders were in place. However, it was only the first step in transitioning from futures and executors to senders, receivers, and schedulers. In this issue I'll try to collect a rough outline of what needs to be done. More detailed discussions are probably better put into separate issues, but high-level discussion is good to keep here. Feel free to modify the lists below. The stretch goals go slightly outside of just sender support.

First milestone:
- [x] #464
- [x] #465
- [x] #466
- [x] #467
- [x] Add a `cuda_scheduler` to HPX. This needs to include the CUDA/cuBLAS/cuSOLVER pools in some form (https://github.com/pika-org/pika/pull/37).

Second milestone:
- [ ] #468

Short and long term stretch goals, i.e. unnecessary for the core of DLA-Future, but very nice to have from a design perspective:
- [ ] Make the tile operations customization points. They should be customizable based on the completion scheduler of the predecessor sender. This should likely use `tag_invoke` for customization like the HPX customization points.
- [ ] Customize the tile operations for `thread_pool_scheduler` and `cuda_scheduler`.
- [ ] Align the tile operations with https://wg21.link/p1673 (note: only BLAS, but LAPACK extensions could likely fit easily).
- [ ] Make the tile operations support `mdspan` arguments instead of only `Tile`s.
- [ ] Put the tile operations (at least the BLAS/LAPACK stuff) in a standalone library.
- [ ] Put the MPI sender functionality in a standalone library.

Additional notes:
- The "API cleanup/updates" points all already align things closer towards proposals. @albestro note that there is no mention of e.g. `dlaf::transform` there ;) Something like it would be an implementation detail of the tile operation customizations. 
- The "Stretch goals" are more of a conformance thing. I think they are the correct way to go long-term. However, since they are not essential for the high-level DLAF API they are likely lower priority.
- I'm treating HIP as equivalent to CUDA.
- The stretch goals imply three major "components" in DLAF: 1. on-node tile operations, aligned as closely as possible with sender/receiver and BLAS proposals, 2. sender/receiver support for MPI, 3. Distributed linear algebra built on top of 1. and 2.
- Each of the lists are ordered roughly according to in which order the tasks should be done.

The first "Basics" task is more or less done, but I think there are still some missing tile operations that have not been senderified. The first "API cleanup/updates" task is something that I've talked about doing for a long time without anything happening, and I think it's a good next step that's orthogonal to the "Basics" tasks.

**Comment by msimberg** [comment]
I think this issue is still useful for the stretch goals, but for the first and second milestone the actual milestones should be referred to for tracking. This issue is somewhat redundant for them.

**Comment by msimberg** [comment]
I'm closing this as most of the points have been made into separate issues and/or milestones.
